tokio/src/sync/batch_semaphore.rs

```diff
@@ -123,6 +123,21 @@ impl Semaphore {
         }
     }
 
+    /// Creates a new semaphore with the initial number of permits
+    ///
+    /// Maximum number of permits on 32-bit platforms is `1<<29`.
+    #[cfg(all(feature = "parking_lot", not(all(loom, test)),))]
+    pub(crate) const fn const_new(permits: usize) -> Self {
+        // FIXME: assertions and by extension panics are still being worked on: https://github.com/rust-lang/rust/issues/74925
```

Comment:
Can the missing assertion cause trouble? E.g. do we have unsafe code that relies on it?

Reply:
Any code that would've been relying on the assertion would have already been using the alternate `new` constructor, so maybe a note in the doc-string about safety is warranted here?

Reply:
tldr; no I don't think there's anything that would cause trouble here

Reply:
> Any code that would've been relying on the assertion would have already been using the alternate new constructor.

That's not really what I mean. We now have a safe function that allows creation of a semaphore with too many permits. Does that make the API unsound, i.e. does it expose a way of using it that causes UB using only the safe interface?

For example, let's say someone comes along and adds a `const fn new` to `Semaphore`. Could that cause undefined behavior?

Reply:
I don't know enough about how the permits function or what they're used for to answer that unfortunately, perhaps someone else can chime in?

Reply:
We do need to ensure `permits` does not exceed the max number of permits for it to be safe. We can do this by saturating probably. If `permits` is greater than `MAX_PERMITS`, then limit it to `MAX_PERMITS`. This can be documented by the fn.

Reply:
I made a change where a guard was introduced to clamp `permits` to `MAX_PERMITS` when exceeded. I did not realize however that if's aren't allowed in const fn's on stable yet so this change is now moot.

Should I mark the function `unsafe` and document the requirements under a `safety` section in the doc-comment instead? @carllerche

Reply:
I've investigated it a little more after observing the other tests passing and it looks like [this test](https://github.com/tokio-rs/tokio/pull/2790/checks?check_run_id=1089992522) is an outlier with an out-of-date stable version `1.45.2` where the const-fn's can't have if's in them but as of stable `1.46.0` they can.

This raises the question, should the test be fixed to use the newer stable version and this usage of this feature now renders a minimal supported rust version of `1.46.0` or should an alternative be applied like described above with marking the constructor `unsafe` and not require an update ?

Reply:
You could do the following since `MAX_PERMITS` is of the form `2^n - 1`.
```
permits = permits & Self::MAX_PERMITS;
```